### PR TITLE
Fix failing relative imports in __init__ files.

### LINKED ISSUE #3442: `from . cimport modulename` tries to search in parent package from `__init__.pxd`.
In `__init__.pxd` in a package, using `from . cimport modulename` attempts to find `modulename.pxd` in the parent directory of the package.

Using `from . import modulename` in basic Python attempts to find `modulename.*` within the package itself, as does using `from . cimport modulename` in all `*.pxd` files *except* for `__init__.pxd` in packages.

Using `from .packagename cimport modulename` in `__init__.pxd` doesn't work around this, because this (correctly) attempts to find `packagename` inside the package itself.

Because of this, I think there should not be much functional code out there that depends on the relative `cimport` being based on the parent directory, meaning that it should be possible to switch it to the package directory, in line with Python behaviour, without breaking compatibility.

---

#### Example structure:
```
cyrel_package/
├── __init__.pxd
├── __init__.py
├── module.pxd
├── module.py
└── subpackage
    ├── __init__.pxd
    ├── __init__.py
    ├── submodule.pxd
    └── submodule.py
```

And in `cyrel_package/subpackage/__init__.pxd`:

```python3
# from . cimport submodule
# Fails, "'cyrel_package/submodule.pxd' not found" — Attempts to import relative to parent package directory instead of current package directory.

# from .subpackage cimport submodule
# Fails, "'cyrel_package/subpackage/subpackage/submodule.pxd' not found" — Attempts to (correctly) import relative to current package directory.

from cyrel_package.subpackage cimport submodule
# Works.
```

It also fails with `from . cimport module` in the topmost `cyrel_package/__init__.pxd`.

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes #3442

### CODE REVIEW

**Comment by matusvalo** [comment]
I have investigated also the cause of #4552. It seems that `analyse_declarations()` is called for following ModuleScopes:

* `PKG`
* `PKG.__init__`

In our case it is `pkg.sub` and `pkg.sub.__init__`:

https://github.com/cython/cython/blob/db787a453996d7441f5791dd901b284b4cfbbfb9/Cython/Compiler/Nodes.py#L8568-L8575

 The main problem is module scope `pkg.sub` which is causing crashing `find_module()` called from `analyse_declaration()` above. When `find_module()` is called for`pkg.sub`, it must not go into following loop.
  https://github.com/cython/cython/blob/db787a453996d7441f5791dd901b284b4cfbbfb9/Cython/Compiler/Symtab.py#L1365-L1367

Hence the main problem is to determine the module scopes of packages (not ending with `__init__` module) and avoid them entering the loop above.

For example, this TC can be easily fixed by as follows (of course this is not generic fix):

```python
            while relative_level > 0 and relative_to and self.name != 'sub':
                relative_to = relative_to.parent_module
                relative_level -= 1
```

Currently, I am stuck how can I determine which ModuleScope is representing package and which module.

**Comment by da-woods** [comment]
> Currently, I am stuck how can I determine which ModuleScope is representing package and which module. I would like to proceed with fix, but I think I will need a hint how :-).

I could be wrong (I've only had a quick look through) but: `ModuleScope` has an `is_package` attribute. It doesn't look to be set correctly for `pxd` modules. I would think that's where you should start. Maybe put a breakpoint in `ModuleScope.__init__` and see if you can work out where scopes are being created that should be packages?

**Comment by matusvalo** [comment]
> I could be wrong (I've only had a quick look through) but: ModuleScope has an is_package attribute.

Yes I saw that. But it is set only for `__init__` ModuleScopes - in our case `<ModuleScope pkg.sub.__init__>`. Is it OK to set `is_package` attribute to `True` also for `<ModuleScope pkg.sub>` (it make sense since `pkg.sub` is a package)?
 
https://github.com/cython/cython/blob/db787a453996d7441f5791dd901b284b4cfbbfb9/Cython/Compiler/Symtab.py#L1240-L1252

**Comment by da-woods** [comment]
>  Is it OK to set `is_package` attribute to True also for `<ModuleScope pkg.sub>`

I think so. It's definitely the first thing I'd try if I were trying to fix it myself.

**Comment by matusvalo** [comment]
This is the first attempt to fix the issue. On my local computer it fixes the TC, I hope I did not introduce any regression. I have tried to set as `is_package` for all `ModuleScope`s of packages (as proposed here: https://github.com/cython/cython/pull/4552#issuecomment-1004063493) but it seems it breaks Cython. Moreover, It was difficult to distinguish ModuleScope of package with `__init__` and without. In the end, I did the most straightforward solution - ModuleScope of package must be one which contains `__init__.py` file which is pretty close to definition of package itself.

Please do a review of the code. I am open to any improvements of the solution.

File: Cython/Compiler/Symtab.py [hunk 1]
```diff
@@ -1351,6 +1352,17 @@ def declare_builtin(self, name, pos):
         entry.qualified_name = self.builtin_scope().qualify_name(name)
         return entry
 
+    def _is_package_scope_or_module(self):
+        # Returns True for all ModuleScopes representing package or scopes representing
+        # modules. Otherwise returns False.
+        # Note: For package pkg_a Cython creates two modules scopes: pkg_a and pkg_a.__init__.
```
Comment (scoder) [inline Cython/Compiler/Symtab.py:1358]: I wonder if it's even correct that we create two separate module scopes. And a good thing. It seems to me that the answer might be no.

File: Cython/Compiler/Symtab.py [hunk 2]
```diff
@@ -1362,7 +1374,7 @@ def find_module(self, module_name, pos, relative_level=-1):
             # explicit relative cimport
             # error of going beyond top-level is handled in cimport node
             relative_to = self
-            while relative_level > 0 and relative_to:
+            while relative_level > 0 and relative_to and self._is_package_scope_or_module():
```
Comment (scoder) [inline Cython/Compiler/Symtab.py:1377]: This will probably be reworked, but it seems wasteful to do something as expensive as this as part of a loop, even though it's obvious that the result will never change. Which, BTW, seems problematic. My guess is that you'd want to check parent packages as well, not just the package directory itself.

**Comment by matusvalo** [comment]
Added additional TC of relative imports in __init__.py. This TC will again crash CI. Fix is easy, removing following lines:
https://github.com/cython/cython/blob/db787a453996d7441f5791dd901b284b4cfbbfb9/Cython/Compiler/Nodes.py#L8572-L8574
Is there any purpose of this error?

File: Cython/Compiler/Nodes.py [hunk 3]
```diff
@@ -8569,9 +8569,6 @@ def analyse_declarations(self, env):
         if not env.is_module_scope:
             error(self.pos, "cimport only allowed at module level")
             return
-        if self.relative_level and self.relative_level > env.qualified_name.count('.'):
-            error(self.pos, "relative cimport beyond main package is not allowed")
-            return
```
Comment (matusvalo) [inline Cython/Compiler/Nodes.py:8574]: This part was removed, in order to fix failing TC introduced in: https://github.com/cython/cython/pull/4552/commits/bd361e215277f81d1ac358269adf34879007e804

File: tests/errors/e_relative_cimport.pyx [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-from ..cython cimport declare
-from . cimport e_relative_cimport
-
-
-_ERRORS="""
-4:0: relative cimport beyond main package is not allowed
-5:0: relative cimport beyond main package is not allowed
-6:0: relative cimport beyond main package is not allowed
-7:0: relative cimport beyond main package is not allowed
-"""
```
Comment (matusvalo) [inline tests/errors/e_relative_cimport.pyx:15]: Tests are removed due removed error here:
https://github.com/cython/cython/pull/4552/files#r779214022

File: Cython/Compiler/Main.py [hunk 5]
```diff
@@ -182,16 +197,16 @@ def find_module(self, module_name, relative_to=None, pos=None, need_pxd=1,
             if not scope:
                 pxd_pathname = self.find_pxd_file(qualified_name, pos)
                 if pxd_pathname:
-                    scope = relative_to.find_submodule(module_name)
+                    is_package = True if '__init__' in pxd_pathname else False
```
Comment (matusvalo) [inline Cython/Compiler/Main.py:200]: This is slightly simple and stupid test just for now. Maybe we can improve it.

**Comment by matusvalo** [comment]
Tests `cpdef_enums` and `cpdef_enums_import` failed in some jobs, but both are executed correctly on my local machine. I am not sure why they failed in CI :-(

File: Cython/Compiler/Symtab.py (hunk 1 above)
Comment (matusvalo) [reply Cython/Compiler/Symtab.py]: Yeas I agree. I fixed in in next version of this PR - see below.

File: Cython/Compiler/Symtab.py (hunk 2 above)
Comment (matusvalo) [reply Cython/Compiler/Symtab.py]: Also reworked in next version of PR.

**Comment by matusvalo** [comment]
I have done diff of generated C __init__ files between latest commit of this PR and commit https://github.com/cython/cython/pull/4552/commits/4a2e6df31167b08faac38317cd5766fb130131cb (basically before removing `PKG.__init__` Module Scope and after removing it). I have found out that there are some differences between generated C files. I am not sure whether the changes can cause issues or not. I would like to ask you to check whether it is fine or not:

```diff
$ diff -u pkg/__init__.c pkg/__init__.c_full
--- pkg/__init__.c	2022-01-06 01:07:54.000000000 +0100
+++ pkg/__init__.c_full	2022-01-04 14:21:31.000000000 +0100
@@ -943,8 +943,8 @@
   #endif
 #endif

-#define __PYX_HAVE__pkg
-#define __PYX_HAVE_API__pkg
+#define __PYX_HAVE__pkg____init__
+#define __PYX_HAVE_API__pkg____init__
 /* Early includes */
 #ifdef _OPENMP
 #include <omp.h>
@@ -1423,7 +1423,7 @@

 /* #### Code section: module_declarations ### */

-/* Module declarations from "pkg" */
+/* Module declarations from "pkg.__init__" */
 #if !CYTHON_USE_MODULE_STATE
 #endif
 /* #### Code section: typeinfo ### */
@@ -1432,7 +1432,7 @@
 extern int __pyx_module_is_main_pkg____init__;
 int __pyx_module_is_main_pkg____init__ = 0;

-/* Implementation of "pkg" */
+/* Implementation of "pkg.__init__" */
 /* #### Code section: global_var ### */
 /* #### Code section: string_decls ### */
 static const char __pyx_k_[] = "?";
@@ -2000,13 +2000,13 @@
   __Pyx_XDECREF(__pyx_t_2);
   if (__pyx_m) {
     if (__pyx_d && stringtab_initialized) {
-      __Pyx_AddTraceback("init pkg", __pyx_clineno, __pyx_lineno, __pyx_filename);
+      __Pyx_AddTraceback("init pkg.__init__", __pyx_clineno, __pyx_lineno, __pyx_filename);
     }
     #if !CYTHON_USE_MODULE_STATE
     Py_CLEAR(__pyx_m);
     #endif
   } else if (!PyErr_Occurred()) {
-    PyErr_SetString(PyExc_ImportError, "init pkg");
+    PyErr_SetString(PyExc_ImportError, "init pkg.__init__");
   }
   __pyx_L0:;
   __Pyx_RefNannyFinishContext();
```

**Comment by matusvalo** [comment]
@scoder @da-woods could you please have a look on the last changes in this PR? Do you have any comments? I would like to understand whether my direction is fine or my approach needs to be changed.

**Comment by scoder** [comment]
> Tests `cpdef_enums` and `cpdef_enums_import` failed in some jobs, but both are executed correctly on my local machine. I am not sure why they failed in CI :-(

Sorry, my fault. I introduced a bug in master. You can try again now.

**Comment by scoder** [comment]
Regarding the module scopes, I tried the following:
```
>>> import Cython
>>> Cython.__file__
'Cython/__init__.py'
>>> Cython.__name__
'Cython'
>>> import Cython.__init__
>>> Cython.__init__.__name__
'Cython.__init__'
>>> Cython.__init__.__file__
'Cython/__init__.py'
```
These attributes are set at import time, so I don't know if it makes a difference for us, but I think we need to make sure that we still follow Python's behaviour here. Basically, whether the package or the module gets imported does make a difference, and I'm not sure if we need to consider that difference for cimports etc.

File: Cython/Compiler/Main.py [hunk 6]
```diff
@@ -143,6 +143,21 @@ def process_pxd(self, source_desc, scope, module_name):
...
+        # Splits qualified_name into parts in form of 2-tuples: (PART_NAME, IS_PACKAGE).
+        qualified_name_parts = qualified_name.split('.')
+        last_part = qualified_name_parts[-1]
+        qualified_name_parts = [(p, True) for p in qualified_name_parts[:-1]]
+        if last_part != '__init__':
+            # If Last part is __init__, then it is omitted. Otherwise, we need to check whether we can find
+            # __init__.pyx/__init__.py file to determine if last part is package or not.
+            path = self.search_include_directories(qualified_name, suffix='.pyx')
+            if not path:
+                path = self.search_include_directories(qualified_name, suffix='.py')
```
Comment (scoder) [inline Cython/Compiler/Main.py:156]: This duplication seems wasteful. And it's also not what (I think) we do normally. I think we rather prefer the first that we find, regardless of the type. There probably code for that already.

File: Cython/Compiler/Nodes.py (hunk 3 above)
Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: I have fixed the condition and introduced back this error.

File: tests/errors/e_relative_cimport.pyx (hunk 4 above)
Comment (matusvalo) [reply tests/errors/e_relative_cimport.pyx]: This test was introduced back in reduced form.

File: Cython/Compiler/Main.py (hunk 6 above)
Comment (matusvalo) [reply Cython/Compiler/Main.py]: I have created new method for that. I was not able to find any existing method doing this.

**Comment by matusvalo** [comment]
The comments were implemented. Ready for next round of review.

File: Cython/Compiler/Nodes.py [hunk 7]
```diff
@@ -8569,7 +8569,7 @@ def analyse_declarations(self, env):
         if not env.is_module_scope:
             error(self.pos, "cimport only allowed at module level")
             return
-        if self.relative_level and self.relative_level > env.qualified_name.count('.'):
+        if self.relative_level and self.relative_level > len(env.qualified_name.split('.')):
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:8572]: This is essentially just `count('.') + 1`. Why is this needed?

File: tests/errors/e_relative_cimport.pyx [hunk 8]
```diff
@@ -2,14 +2,10 @@
...
 from ..cython cimport declare
-from . cimport e_relative_cimport
 
 
 _ERRORS="""
 4:0: relative cimport beyond main package is not allowed
 5:0: relative cimport beyond main package is not allowed
-6:0: relative cimport beyond main package is not allowed
-7:0: relative cimport beyond main package is not allowed
 """
```
Comment (scoder) [inline tests/errors/e_relative_cimport.pyx:11]: These errors look correct for a top-level module file. Relative imports should only be allowed inside of packages.

**Review by scoder — COMMENTED** [review]
I left a couple of comments.

Could you make sure that we compare the behaviour of normal (uncompiled) Python to the behaviour that we create here? You could add a plain Python package and/or sub-package to the srctree tests and then import the Python package and the Cython package, and do your assertions on both to compare.

File: Cython/Compiler/Nodes.py (hunk 7 above)
Comment (matusvalo) [reply Cython/Compiler/Nodes.py:8572]: When this change is not present following case is failing:

Suppose we have following package:
```
pkg/
  + __init__.pyx
  + __init__pxd
  + sub/
     + __init__.pyx
     + __init__.pxd
```
And `pkg/__init__.pyx` contains following line:

```cython
from . cimport sub
```

Command `cythonize -3ib __init__.pyx` fails with following:
```
Compiling /Users/matus/dev/cython/test/pkg/__init__.pyx because it changed.
[1/1] Cythonizing /Users/matus/dev/cython/test/pkg/__init__.pyx
> /Users/matus/dev/cython/Cython/Compiler/Nodes.py(8576)analyse_declarations()
-> error(self.pos, "relative cimport beyond main package is not allowed")
(Pdb) c

Error compiling Cython file:
------------------------------------------------------------
...
from . cimport sub
^
------------------------------------------------------------

__init__.pyx:1:0: relative cimport beyond main package is not allowed
Traceback (most recent call last):
  File "/Users/matus/dev/cython39/bin/cythonize", line 33, in <module>
    sys.exit(load_entry_point('Cython', 'console_scripts', 'cythonize')())
  File "/Users/matus/dev/cython/Cython/Build/Cythonize.py", line 224, in main
    cython_compile(path, options)
  File "/Users/matus/dev/cython/Cython/Build/Cythonize.py", line 68, in cython_compile
    ext_modules = cythonize(
  File "/Users/matus/dev/cython/Cython/Build/Dependencies.py", line 1117, in cythonize
    cythonize_one(*args)
  File "/Users/matus/dev/cython/Cython/Build/Dependencies.py", line 1284, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: /Users/matus/dev/cython/test/pkg/__init__.pyx
```

Comment (matusvalo) [reply Cython/Compiler/Nodes.py:8572]: It is failing because in this case `self.relative_level` is `1` and `env.qualified_name` is `'pkg'`

**Comment by matusvalo** [comment]
> Could you make sure that we compare the behaviour of normal (uncompiled) Python to the behaviour that we create here? You could add a plain Python package and/or sub-package to the srctree tests and then import the Python package and the Cython package, and do your assertions on both to compare.

@scoder  I have implemented unittests. Please have a look if you are fine with that.

File: tests/errors/e_relative_cimport.pyx (hunk 8 above)
Comment (matusvalo) [reply tests/errors/e_relative_cimport.pyx]: The errors are fixed.

**Comment by matusvalo** [comment]
@scoder @da-woods I have implemented comments. Could you review the PR?

File: Cython/Compiler/Nodes.py (hunk 7 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py]: > When this change is not present following case is failing:

Could you have another double-check on this change. @scoder was suggesting adding `+ 1` to what's currently there, which I think should fix your `from . cimport sub` example

File: Cython/Compiler/Main.py [hunk 9]
```diff
@@ -143,6 +143,23 @@ def process_pxd(self, source_desc, scope, module_name):
     def nonfatal_error(self, exc):
         return Errors.report_error(exc)
 
+    def _split_qualified_name(self, qualified_name):
+        # Splits qualified_name into parts in form of 2-tuples: (PART_NAME, IS_PACKAGE).
+        qualified_name_parts = qualified_name.split('.')
+        last_part = qualified_name_parts.pop()
```
Comment (da-woods) [inline Cython/Compiler/Main.py:149]:
```suggestion
        *qualified_name_parts, last_part = qualified_name.split('.')
```

**Review by da-woods — COMMENTED** [review]
Just a few small style suggestions from me...

File: Cython/Compiler/Nodes.py (hunk 7 above)
Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: I missed this comment. I will check that.

Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: I have reverted the change back to `count('.') + 1`

**Comment by matusvalo** [comment]
@scoder @da-woods I think all comments were fixed. Could you have a look on the PR?

**Comment by matusvalo** [comment]
@scoder is this PR able to make it the next release? It would be great if yes... Let me know if there is anythong pending. I will try to fix it asap.

**Comment by matusvalo** [comment]
Based on comment https://github.com/cython/cython/issues/4022#issuecomment-1200386975 can be this PR merged before beta of 3.0.0? (link back to https://github.com/cython/cython/issues/4022)

File: Cython/Compiler/Symtab.py [hunk 10]
```diff
@@ -1421,9 +1415,18 @@ def find_module(self, module_name, pos, relative_level=-1):
...
+                # i.e. dot in `from . import ...` points to current package
+                while relative_level > 1 and relative_to:
+                    relative_to = relative_to.parent_module
+                    relative_level -= 1
+            else:
+                # When file is regular module, current package is parent module
+                # i.e. dot in `from . import ...` points to package where module is placed
+                while relative_level > 0 and relative_to:
+                    relative_to = relative_to.parent_module
+                    relative_level -= 1
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Compiler/Symtab.py]: fixed. Comments are kept.

File: Cython/Compiler/Nodes.py [hunk 11]
```diff
@@ -8728,9 +8728,14 @@ def analyse_declarations(self, env):
...
-        if self.relative_level and self.relative_level > env.qualified_name.count('.'):
-            error(self.pos, "relative cimport beyond main package is not allowed")
-            return
+        qualified_name_components = env.qualified_name.count('.') + 1
+        if self.relative_level:
+            if self.relative_level > qualified_name_components or (
+                    self.relative_level == qualified_name_components and not env.is_package):
+                # 1. case: importing beyond package: from .. import pkg
+                # 2. case: importing from same level but current dir is not package: from . import module
+                error(self.pos, "relative cimport beyond main package is not allowed")
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Compiler/Nodes.py]: fixed.

File: Cython/Compiler/Main.py [hunk 12]
```diff
@@ -254,6 +271,15 @@ def find_pyx_file(self, qualified_name, pos=None, source_file_path=None):
...
             qualified_name, suffix=".pyx", source_pos=pos, source_file_path=source_file_path)
 
+    def find_file_with_suffix(self, qualified_name, pos=None, source_file_path=None, suffix=('.py', '.pyx', '.pxd')):
+        found = []
+        for s in suffix:
+            path = self.search_include_directories(
+                qualified_name, suffix=s, source_pos=pos, source_file_path=source_file_path)
+            if path:
+                found.append(path)
+        return found
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Compiler/Main.py]: done.

File: tests/run/relative_cimport_compare.srctree [hunk 13]
```diff
@@ -0,0 +1,327 @@
+# mode: run
+# tag: cimport, pep489
+
+PYTHON setup.py build_ext --inplace
+PYTHON -c "import test_import"
+PYTHON -c "import test_cimport"
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply tests/run/relative_cimport_compare.srctree:6]: > They look generated

Unfortunately not, they are manually created.

> somehow use code generation for the package directory structure and test files?

Is there anywhere a way how to do it or I need to do it from scratch?

[the remark this answers is not in the document]
Comment (da-woods) [reply tests/run/relative_cimport_compare.srctree:6]: > Is there anywhere a way how to do it or I need to do it from scratch?

We don't have a specific framework to do this kind of thing

If it was generated then I'd probably suggest either:
* including the generation script in the commit (maybe Tools/), or
* putting the script as part of a srctree file and run it as step 1 (instead of including the output in the srctree file)

Given they aren't generated I don't think this requires a change. It would just have been a shame not to have the script if it already existed.

[the remark this answers is not in the document]
Comment (matusvalo) [reply tests/run/relative_cimport_compare.srctree:6]: No unfortunatelly I don't have such tool... So I am resolving the conversion. Please let me know if there is anything blocking this PR being accepted.

**Comment by matusvalo** [comment]
The CI is failed because windows build was cancelled after 50 minutes :-(

**Comment by matusvalo** [comment]
@da-woods @scoder is there anything pending here? All comments were implemented.

**Comment by da-woods** [comment]
Thanks @matusvalo

**Comment by matusvalo** [comment]
Thank you @da-woods for merging this PR!
